### PR TITLE
Tests for reward trainers improving reward loss

### PR DESCRIPTION
**Pull request by dan-pandori**
## Description

Small test ensuring that BasicRewardTrainer and EnsembleRewardTrainer's train functions leads to increased reward/lower losses.

https://github.com/HumanCompatibleAI/imitation/issues/562

## Testing

I ran this with flakefinder to find a lowish amount of training time that would still have a low flake rate, and confirmed that removing the loss.backward calls in BasicRewardTrainer's _train method consistently caused the test to fail.

### CODE REVIEW

**Comment by dan-pandori** [comment]
@AdamGleave to take a look when you've got a chance.

The failing test is from the other type issue.

File: tests/algorithms/test_preference_comparisons.py [hunk 1]
```diff
@@ -918,3 +920,141 @@ def test_agent_trainer_sample_image_observations(rng):
...
         for trajectory in trajectories
     )
+
+
+class ActionIsRewardEnv(gym.Env):
+    """Two step environment where the reward is the sum of actions."""
+
+    def __init__(self):
+        """Initialize environment."""
+        super(ActionIsRewardEnv, self).__init__()
```
[the remark this answers is not in the document]
Comment (dan-pandori) [reply tests/algorithms/test_preference_comparisons.py:930]: Done, thanks!

File: tests/algorithms/test_preference_comparisons.py [hunk 2]
```diff
@@ -918,3 +920,141 @@ def test_agent_trainer_sample_image_observations(rng):
...
+        self.observation_space = gym.spaces.Box(np.array([0]), np.array([1]))
+        self.reward = 0.0
+        self.steps = 0
+
+    def step(self, action):
+        obs = np.array([0])
+        reward = action
+        # Some algorithms expect trajectories at least two steps long,
+        # so we allow two steps to be taken.
+        done = self.steps > 0
```
[the remark this answers is not in the document]
Comment (dan-pandori) [reply tests/algorithms/test_preference_comparisons.py:941]: +, good to know for future reference

File: tests/algorithms/test_preference_comparisons.py [hunk 3]
```diff
@@ -918,3 +920,141 @@ def test_agent_trainer_sample_image_observations(rng):
...
+
+    def reset(self):
+        self.steps = 0.0
+        return np.array([0])
+
+
+@pytest.fixture
+def action_is_reward_venv(rng):
+    return DummyVecEnv(
+        [partial(ActionIsRewardEnv)],
```
[the remark this answers is not in the document]
Comment (dan-pandori) [reply tests/algorithms/test_preference_comparisons.py:954]: Nope, I was copy pasting w/o thinking it through completely. Thanks for the cleanup.

File: tests/algorithms/test_preference_comparisons.py [hunk 4]
```diff
@@ -918,3 +920,141 @@ def test_agent_trainer_sample_image_observations(rng):
...
+        self.steps += 1
+        return obs, reward, done, info
+
+    def reset(self):
+        self.steps = 0.0
+        return np.array([0])
+
+
+@pytest.fixture
+def action_is_reward_venv(rng):
```
[the remark this answers is not in the document]
Comment (dan-pandori) [reply tests/algorithms/test_preference_comparisons.py:952]: Good point, removed

File: tests/algorithms/test_preference_comparisons.py [hunk 5]
```diff
@@ -918,3 +920,141 @@ def test_agent_trainer_sample_image_observations(rng):
...
+        rng=rng,
+        reward_trainer=action_is_reward_trainer,
+        custom_logger=custom_logger,
+    )
+
+    # Train the net for a short period of time, and then again.
+    # We expect it to have improved over that period.
+    first_rewards = main_trainer.train(20, 20)
+    later_rewards = main_trainer.train(20, 20)
+    assert first_rewards["reward_loss"] > later_rewards["reward_loss"]
```
[the remark this answers is not in the document]
Comment (dan-pandori) [reply tests/algorithms/test_preference_comparisons.py:1060]: Yep, this is intended but unclear. I added a comment here and at the top of the test.

Also added a TODO at the bottom for adding another test checking that agent performance improves after training.

File: tests/algorithms/test_preference_comparisons.py [hunk 6]
```diff
@@ -918,3 +920,141 @@ def test_agent_trainer_sample_image_observations(rng):
...
+
+class ActionIsRewardEnv(gym.Env):
+    """Two step environment where the reward is the sum of actions."""
+
+    def __init__(self):
+        """Initialize environment."""
+        super(ActionIsRewardEnv, self).__init__()
+        self.action_space = spaces.Discrete(50)
+        self.observation_space = gym.spaces.Box(np.array([0]), np.array([1]))
+        self.reward = 0.0
```
[the remark this answers is not in the document]
Comment (dan-pandori) [reply tests/algorithms/test_preference_comparisons.py:933]: Yep, thanks

File: tests/algorithms/test_preference_comparisons.py [hunk 7]
```diff
@@ -918,3 +920,141 @@ def test_agent_trainer_sample_image_observations(rng):
...
+        "MlpPolicy",
+        action_is_reward_venv,
+        n_epochs=1,
+        batch_size=2,
+        n_steps=10,
+    )
+
+
+@pytest.fixture
+def action_is_reward_trainer(
```
[the remark this answers is not in the document]
Comment (dan-pandori) [reply tests/algorithms/test_preference_comparisons.py:970]: Fair point, I went with just splitting it into two separate helper functions because I feel that using the closure is getting a little fancy to follow.

Thoughts on it now?

**Comment by dan-pandori** [comment]
@AdamGleave, updated from the comments, thanks for the review!

File: tests/algorithms/test_preference_comparisons.py [hunk 8]
```diff
@@ -966,74 +964,73 @@ def action_is_reward_agent(action_is_reward_venv, rng):
     )
 
 
-@pytest.fixture
-def action_is_reward_trainer(
-    action_is_reward_venv,
+def basic_reward_trainer(
```
Comment (AdamGleave) [inline tests/algorithms/test_preference_comparisons.py:967]: Nit: this probably doesn't need to be spread over multiple lines now there's only a single parameter. (black won't put things back together on one line once they have a trailing comma, but will happily split them up when they're too long.)

**Review by AdamGleave — APPROVED** [review]
LGTM (one cosmetic improvement you could make)

File: tests/algorithms/test_preference_comparisons.py (hunk 8 above)
Comment (dan-pandori) [reply tests/algorithms/test_preference_comparisons.py]: Sure, done
